sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,477 @@
+        """
+        return tuple(sorted(request)) in self._denied
+
+    # register, request, resume, acknowledge, deny, etc.
+
+    def register(self, plugin_name: str, request: capability) -> None:
+        """Register a capability ``request`` for ``plugin_name``.
+
+        :param request: the capability request to register for later
+        :raise RuntimeError: when the capability request is too long for a
+                             single ``CAP REQ`` and ``CAP * ACK``
+
+        Once registered, the capability request can be requested by the bot. A
+        registered request appears in :attr:`registered`::
+
+            >>> from sopel import plugin
+            >>> request = plugin.capability('cap1')
+            >>> manager.register('coretasks', request)
+            >>> ('cap1',) in manager.registered
+            True
+            >>> manager.is_registered(('cap1',))
+            True
+
+        It is not, however, directly requested::
+
+            >>> manager.is_requested(('cap1',))
+            False
+
+        See :meth:`request_available` to automatically request capabilities
+        advertised by the server.
+
+        .. warning::
+
+            Sopel cannot accept a request that is too long, because it does not
+            know how to handle a multi-line ACK, and it would not know how to
+            call back the appropriate :class:`~sopel.plugin.capability`
+            handler.
+
+        """
+        cap_req = ' '.join(request.cap_req)
+        if len(cap_req.encode('utf-8')) > 500:
+            # "CAP * ACK " is 10 bytes, leaving 500 bytes for the capabilities.
+            # Sopel cannot allow multi-line requests, as it won't know how to
+            # deal properly with multi-line ACK.
+            # The spec says a client SHOULD send multiple requests; however
+            # the spec also says that a server will ACK or NAK a whole request
+            # at once. So technically, multiple REQs are not the same as a
+            # single REQ.
+            raise RuntimeError('Capability request too long: %s' % cap_req)
+
+        plugin_caps = self._registered.setdefault(request.cap_req, {})
+        plugin_caps[plugin_name] = (
+            request, False,
+        )
+        LOGGER.debug('Capability Request registered: %s', str(request))
+
+    def request_available(
+        self,
+        bot: Sopel,
+        available_capabilities: Iterable[str],
+    ) -> None:
+        """Request available capabilities.
+
+        :param bot: the bot instance used to send capability requests
+        :param available_capabilities: available capabilities
+
+        This sends ``CAP REQ`` commands for requests that can be made, i.e.
+        all the requested capabilities (with or without prefix) must be
+        available for Sopel to send the request.
+
+        Requests made are stored as requested; others are ignored::
+
+            >>> # manager.register => cap1 and cap2, not cap3
```

Comment:
I just realized I don't understand this example. What is this line of the repr supposed to represent? Since the `register()` method above doesn't return anything, I think that's where I'm getting lost.

Reply:
That will make more sense.

```suggestion
            >>> manager.register('example', 'cap1')
            >>> manager.register('example', 'cap2')
```